ifm: init at 4.0.2

litchipi opened the pull request:
## Description of changes

Create a web-app module for a single-file service called IFM for Improved File Manager.

See homepage here: https://github.com/misterunknown/ifm

I thought it wasn't worth a package (as it's just downloading a file), so only the module is done here.

Not really familiar with PHP's web server, so I don't know if that's an acceptable way to serve the PHP file, or if we should set up a nginx / caddy / whatever in the module as a replacement.

For security purposes, the systemd service is isolated, and only has a bind mount access to the `dataDir` provided in the configuration.

Didn't test the advanced configuration as described here: https://github.com/misterunknown/ifm/wiki/Configuration, particularly the authentication, but as it's not very "bound" to the operating system (except ldap), it should not break.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

litchipi committed:
Init IFM service at version 4.0.2

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4279

gador reviewed (changes requested):
I think for such a small package it is fine to not add an extra package. Also I tend to aggree to not use `nginx` or `httpd` (or any other webserver) here. You have the benefit of keeping this module small and not pulling in extra dependencies. 
Also some people use `nginx` other `httpd` others `caddy` etc. This way anyone can use their preferred webserver as a reverse proxy to serve `ifm`

litchipi commented on nixos/modules/services/web-apps/ifm.nix:
```diff
@@ -0,0 +1,84 @@
+      description = "Improved file manager, a single-file web based filemanager";
+
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
```
Don't really know the inner workings of systemd services, so I just copied something that worked for `mealie` :sweat_smile: 
I'll remove it and make some tests

litchipi replied:
Doesn't work without it, and even throws an error

gador replied:
```suggestion
```

gador replied:
I just pulled your PR and ran the `nixosTest` with this change and it works as it should. What error did it throw for you?

litchipi replied:
My bad, last time I removed the `wantedBy` line, not the `wants` one :sweat_smile:

litchipi commented on nixos/tests/ifm.nix:
```diff
@@ -0,0 +1,25 @@
+    };
+  };
+
+  testScript = ''
+    start_all()
+    server.wait_for_unit("ifm.service")
+    server.wait_for_open_port(9001)
+    server.succeed("curl --fail http://localhost:9001")
```
Took me a while, but I think I have something satisfying now :smiley:

litchipi commented:
Just saw online that the PHP builtin web server is terrible for production usage, so I'll try to find something else to serve IFM.
But I think the less overhead it brings, the better it'll be

SuperSandro2000 commented on nixos/modules/services/web-apps/ifm.nix:
```diff
@@ -0,0 +1,75 @@
+          mkdir -p /tmp/ifm
+          ln -s ${src} /tmp/ifm/index.php
+        '';
+        ExecStart = "${lib.getExe php} -S ${cfg.listenAddress}:${builtins.toString cfg.port} -t /tmp/ifm";
```
Do we have PrivateTmp set?

litchipi replied:
Woops, thought that was default to `true`, will set it, thanks

litchipi commented:
Made the changes, and pushed a new version.

Experimented a lot of things to see if we could serve it using something else than PHP's built-in server, but it seam overly complicated compared to the scope of the project.
As the Dockerfile of the IFM github repository uses php's builtin server, the app is probably built around single-threading anyway.

So I reverted back to using it, and added a disclaimer in the `enable` description in order to prevent anyone from using it in a heavy prod context.
@gator tell me if that's satisfactory for you :slightly_smiling_face:

gador commented:
> @gator tell me if that's satisfactory for you 🙂

Looks good to me! Just one minor thing with the `systemd` service and then I'd be happy to merge this.
`nixosTests` succeed on `86_64-linux` :+1:

litchipi commented:
@gador Fixed, thanks for the catch

gador reviewed (approved):
Thanks for your work!

pbsds commented on nixos/modules/services/web-apps/ifm.nix:
```diff
@@ -0,0 +1,81 @@
+let
+  cfg = config.services.ifm;
+
+  version = "4.0.2";
+  src = pkgs.fetchurl {
+    url = "https://github.com/misterunknown/ifm/releases/download/v${version}/cdn.ifm.php";
+    hash = "sha256-37WbRM6D7JGmd//06zMhxMGIh8ioY8vRUmxX4OHgqBE=";
+  };
```
This should be a package

pbsds replied:
that will enable r-ryantm bumps, and setting `meta.license` and `meta.sourceProvenance`

gador replied:
I remarked on this issue above (https://github.com/NixOS/nixpkgs/pull/328205#pullrequestreview-2186639824) but concluded it would be fine for such a small package (which is basically just one `php` file). 
However I haven't considered the `meta` fields and automatic package updates by ryan. @litchipi maybe you could do a follow-up PR which separates the package `ifm` from the module `ifm`?

litchipi replied:
Working on it, sorry didn't thought of these problems.

However, the name `ifm` is already taken as a package name (in `pkgs/tools/graphics/ifm`), so I don't really know what to name it...

pbsds reviewed (commented):
@gador The merged commit title does not follow the nixpkgs schema, a squash-merge would've solved that

gador commented:
> @gador The merged commit title does not follow the nixpkgs schema, a squash-merge would've solved that

You're right, I must've overlooked the commit title
